### ISSUE TITLE
Double redirect oddness

### ISSUE THREAD

**Issue by xbony2** [fastily/jwiki#11]
On my wiki, I have a double redirect that looks like this:
```
1. Getting Started (IndustrialCraft)/fr (edit) →‎ IndustrialCraft 2/Guide/fr →‎ Getting Started (IndustrialCraft 2)/fr
```
This is structured as double redirect -> normal redirect -> target.

I have been working on a command in my bot used to fix double redirects. My code looks something like this:
```groovy
wiki.querySpecialPage("DoubleRedirects", -1).each {
	def target = wiki.getPageText(it).replaceAll("#REDIRECT \\[\\[", "").replaceAll("]]", "")
	def targetsTarget = wiki.getPageText(target).replaceAll("#REDIRECT \\[\\[", "").replaceAll("]]", "")
	wiki.edit(target, "testy testy"/*"#REDIRECT [[${targetsTarget}]]"*/, "Fixed double redirect.")
}
```
(This code might be kind of confusing if you don't know Groovy and because I have a habit of compacting things but just for your reference)

My bot made [this edit](https://ftb.gamepedia.com/index.php?title=IndustrialCraft_2/Guide/fr&diff=695631&oldid=656176).

Basically, the `querySpecialPage` method using the `"DoubleRedirects"` special page gives an array of the pages of the "normal redirects" in the double redirects page (in this case "IndustrialCraft 2/Guide/fr") instead of the actual double redirects ("Getting Started (IndustrialCraft)/fr").

**Comment by fastily** [comment]
Hi @xbony2,

I've been researching this, and I suspect this may not be possible, at least not through the the API.  Long story short, the API's `querypages` module only returns whatever the `DoubleRedirects` special page gives it; there is no way to influence/customize this behavior.  

Fortunately, there is a workaround, jwiki can resolve redirects.  The following example works on [testwiki](https://test.wikipedia.org): 
```java
Wiki wiki = new Wiki("test.wikipedia.org");	
System.out.println( MQuery.resolveRedirects(wiki, wiki.querySpecialPage("DoubleRedirects", 3)) );
```

**Comment by xbony2** [comment]
What do you mean by resolve redirects?

**Comment by fastily** [comment]
In this context, 'resolving redirects' is defined as fetching the target of a redirect page.

**Comment by xbony2** [comment]
Sorry, but I'm confused 😕 by "resolving redirects," aren't I already doing that with the `def target = wiki.getPageText(it).replaceAll("#REDIRECT \\[\\[", "").replaceAll("]]", "")` bit? (the keyword "it" is the default name for the current iteration, which yields "IndustrialCraft 2/Guide/fr" for the one iteration).

**Comment by fastily** [comment]
Apologies for the confusion, did some more digging and I get what you mean.   Your Wiki's API returns more information for DoubleRedirects than the WMF Wikis' APIs (i.e. the reference Wiki I've been testing jwiki against).

For example, on your Wiki, this [query](https://ftb.gamepedia.com/Special:ApiSandbox#action=query&format=json&list=querypage&qppage=DoubleRedirects&qplimit=1) (`/api.php?action=query&format=json&list=querypage&qppage=DoubleRedirects&qplimit=1`) returns:

```javascript
{
    "batchcomplete": "",
    "query": {
        "querypage": {
            "name": "DoubleRedirects",
            "results": [
                {
                    "value": "Getting_Started_(IndustrialCraft)/fr",
                    "ns": 0,
                    "title": "Getting Started (IndustrialCraft)/fr",
                    "databaseResult": {
                        "nsb": "0",
                        "tb": "IndustrialCraft_2/Guide/fr",
                        "nsc": "0",
                        "tc": "Getting_Started_(IndustrialCraft_2)/fr",
                        "iwc": ""
                    }
                }
            ]
        }
    }
}
```

On testwiki, [this query](https://test.wikipedia.org/wiki/Special:ApiSandbox#action=query&format=json&list=querypage&qppage=DoubleRedirects&qplimit=1) (`/w/api.php?action=query&format=json&list=querypage&qppage=DoubleRedirects&qplimit=1`) returns

```javascript
{
    "batchcomplete": "",
    "continue": {
        "qpoffset": 1,
        "continue": "-||"
    },
    "query": {
        "querypage": {
            "name": "DoubleRedirects",
            "cached": "",
            "cachedtimestamp": "2018-07-19T13:53:43Z",
            "maxresults": 5000,
            "results": [
                {
                    "value": "0",
                    "ns": 0,
                    "title": "Double2"
                }
            ]
        }
    }
}
```

I don't know why my results don't contain a `databaseResult` object, so I've [asked](https://www.mediawiki.org/wiki/Topic:Uh7qrf2gmxakziw5) about it at the MediaWiki support desk.  I'll update this thread accordingly when I hear back.

**Comment by xbony2** [comment]
okay 👍

**Comment by fastily** [comment]
Quick update: I've [created a task on WMF Phabricator](https://phabricator.wikimedia.org/T200166) to make this a standard part of the API.   Since it may take awhile to happen, I'll explore the feasibility of building a workaround into jwiki.

**Comment by fastily** [comment]
I don't believe the WMF devs will implement server side resolving of double redirects in vanilla MediaWiki anytime soon.  In the interest of keeping the jwiki API as simple as possible, I'm going to forgo a client-side hack.

Here's a workaround for your code.  It's not pretty, but it should get the results you need:

```java
Wiki wiki = new Wiki("test.wikipedia.org");
for(String s : wiki.querySpecialPage("DoubleRedirects", 5))
{
	String doubleRedirectTarget = wiki.resolveRedirect(wiki.resolveRedirect(s));

	// do something with doubleRedirectTarget here...
}
```

**Comment by fastily** [comment]
closing as unactionable.  I'll revisit when MediaWiki implements an API endpoint to get double redirect targets
